hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/BoundedWriteBehindQueue.java

```diff
@@ -0,0 +1,190 @@
+
+    /**
+     * Per node write behind queue item counter.
+     */
+    private final AtomicInteger writeBehindQueueItemCounter;
+
+    /**
+     * Allowed max capacity per node which is used to provide back-pressure.
+     */
+    private final int maxCapacity;
+
+    private final WriteBehindQueue<E> queue;
+
+    BoundedWriteBehindQueue(int maxCapacity, AtomicInteger writeBehindQueueItemCounter, WriteBehindQueue<E> queue) {
+        this.maxCapacity = maxCapacity;
+        this.writeBehindQueueItemCounter = writeBehindQueueItemCounter;
+        this.queue = queue;
+    }
+
+    /**
+     * Inserts to the end of this queue.
+     *
+     * @param e element to be offered
+     */
+    @Override
+    public void addEnd(E e) {
+        ensureRequiredCapacity(1);
+        queue.addEnd(e);
+    }
+
+    /**
+     * Add this collection to the front of the queue.
+     *
+     * @param collection collection of elements to be added in front of this queue.
+     */
+    @Override
+    public void addFront(Collection<E> collection) {
+        if (collection == null || collection.isEmpty()) {
+            return;
+        }
+        ensureRequiredCapacity(collection.size());
+        queue.addFront(collection);
+    }
+
+    /**
+     * Removes the first occurrence of the specified element in this queue
+     * when searching it by starting from the head of this queue.
+     *
+     * @param e element to be removed.
+     * @return <code>true</code> if removed successfully, <code>false</code> otherwise
+     */
+    @Override
+    public boolean removeFirstOccurrence(E e) {
+        boolean result = queue.removeFirstOccurrence(e);
+        if (result) {
+            decrementPerNodeMaxSize(1);
+        }
+        return result;
+    }
+
+    /**
+     * Removes all elements from this queue and adds them
+     * to the given collection.
+     *
+     * @return number of removed items from this queue.
+     */
+    @Override
+    public int drainTo(Collection<E> collection) {
+        final int size = queue.drainTo(collection);
+        decrementPerNodeMaxSize(size);
+        return size;
+    }
+
+    /**
+     * Checks whether an element exist in this queue.
+     *
+     * @param e item to be checked
+     * @return <code>true</code> if exists, <code>false</code> otherwise
+     */
+    @Override
+    public boolean contains(E e) {
+        return queue.contains(e);
+    }
+
+    /**
+     * Returns the number of elements in this {@link WriteBehindQueue}.
+     *
+     * @return the number of elements in this {@link WriteBehindQueue}.
+     */
+    @Override
+    public int size() {
+        return queue.size();
+    }
+
+    /**
+     * Removes all of the elements in this  {@link WriteBehindQueue}
+     * Queue will be empty after this method returns.
+     */
+    @Override
+    public void clear() {
+        final int size = size();
+        queue.clear();
+        decrementPerNodeMaxSize(size);
```

Comment:
I don't know if `BoundedWriteBehindQueue` is modified concurrently but I assume so because of used atomic counter,  then `clear()` method is racy. There's a gap between `queue.size()` call and `queue.clear()`.